File: src/calc_flow/runtime/streaming.py
from __future__ import annotations

from calc_flow.batch import Batch
from calc_flow.checkpoint import (
    CheckpointStore,
    FileCheckpointStore,
    validate_checkpoint,
)
from calc_flow.pipeline import ExecutionPlan, Pipeline, RunResult
from calc_flow.runtime.micro_batch import (
    SinkTarget,
    _coerce_plan,
    _make_checkpoint,
    _single_input_name,
    _write_outputs,
)


class StreamingRunner:
    """Execute one formed batch at a time with at-least-once checkpoints."""

    def __init__(
        self,
        pipeline: Pipeline | ExecutionPlan,
        *,
        checkpoint_store: CheckpointStore | None = None,
    ) -> None:
        self.plan = _coerce_plan(pipeline)
        self.checkpoint_store = (
            checkpoint_store if checkpoint_store is not None else FileCheckpointStore()
        )
        self._input_name = _single_input_name(self.plan)
        self._sequence = 0
        self._recovered = False

    def step(self, batch: Batch, sink: SinkTarget | None = None) -> RunResult:
        """Process, deliver, and checkpoint one batch before returning its result."""
        if not isinstance(batch, Batch):
            msg = "StreamingRunner.step() requires a Batch"
            raise TypeError(msg)
        self._recover_once()
        state_before = self.plan.snapshot()
        result = self.plan.execute({self._input_name: batch})
        try:
            _write_outputs(result, sink)
            checkpoint = _make_checkpoint(
                self.plan,
                cursor=batch.metadata.cursor,
                sequence=self._sequence,
            )
            self.checkpoint_store.save(checkpoint)
        except Exception:
            self.plan.restore(state_before)
            raise
        self._sequence += 1
        return result

    def reset(self) -> None:
        self._sequence = 0
        self.plan.reset()
        self.checkpoint_store.delete(self.plan.name)
        self._recovered = True

    def _recover_once(self) -> None:
        if self._recovered:
            return
        checkpoint = self.checkpoint_store.load(self.plan.name)
        if checkpoint is not None:
            validate_checkpoint(
                checkpoint,
                pipeline_name=self.plan.name,
                fingerprint=self.plan.fingerprint,
            )
            self.plan.restore(checkpoint.state)
            self._sequence = checkpoint.sequence + 1
        self._recovered = True

File: crates/calc-flow/src/runtime/streaming.rs
use std::{collections::BTreeMap, sync::Arc};

use chrono::Utc;

use crate::{
    Batch, CalcFlowError, Checkpoint, CheckpointStore, ExecutionOptions, ExecutionPlan, Result,
    RunResult,
    pipeline::{PlanLease, PlanTransaction},
};

use super::{
    SinkRouter,
    micro_batch::{compensated_reset_error, rollback_error, validate_checkpoint},
};

/// Push-based at-least-once runner for already formed batches.
pub struct StreamingRunner {
    plan: Arc<ExecutionPlan>,
    lease: PlanLease,
    checkpoints: Arc<dyn CheckpointStore>,
    recovered: bool,
    next_sequence: Option<u64>,
    poisoned: Option<String>,
}

impl StreamingRunner {
    /// Constructs a runner for a plan with exactly one external input.
    ///
    /// # Errors
    ///
    /// Returns [`CalcFlowError::InvalidArgument`] unless the plan has exactly
    /// one external input, or [`CalcFlowError::PlanLeased`] when another
    /// runner already owns the plan.
    pub fn new(plan: Arc<ExecutionPlan>, checkpoints: Arc<dyn CheckpointStore>) -> Result<Self> {
        plan.single_external_input()?;
        let lease = plan.acquire_lease()?;
        Ok(Self {
            plan,
            lease,
            checkpoints,
            recovered: false,
            next_sequence: Some(0),
            poisoned: None,
        })
    }

    /// Recovers once, executes and delivers one batch, then checkpoints it.
    ///
    /// Push-mode checkpoints do not own a replay cursor: `source_cursor` is
    /// always `None`. This method consumes the supplied batch. The caller must
    /// retain a clone or reconstruct it, then submit that value again after a
    /// failed step to obtain at-least-once delivery.
    ///
    /// # Errors
    ///
    /// Returns recovery, execution, sink, checkpoint, or rollback errors.
    pub async fn step(&mut self, batch: Batch, sinks: &mut SinkRouter) -> Result<RunResult> {
        self.ensure_healthy()?;
        self.recover_once().await?;
        let sequence = self.next_sequence.ok_or_else(|| CalcFlowError::Internal {
            message: "streaming sequence is exhausted".into(),
        })?;
        let plan = Arc::clone(&self.plan);
        let transaction = plan.leased_transaction(&self.lease).await?;
        let input_name = plan.single_external_input()?.to_owned();
        let inputs = BTreeMap::from([(input_name, batch)]);
        transaction.validate_inputs(&inputs)?;
        let before = transaction.snapshot().await?;
        let result = match transaction
            .execute_validated(inputs, ExecutionOptions::default())
            .await
        {
            Ok(result) => result,
            Err(error) => {
                return Err(self.rollback_operation(&transaction, &before, error).await);
            }
        };
        if let Err(error) = sinks.write_all(&result).await {
            return Err(self.rollback_operation(&transaction, &before, error).await);
        }
        let state = match transaction.snapshot().await {
            Ok(state) => state,
            Err(error) => {
                return Err(self.rollback_operation(&transaction, &before, error).await);
            }
        };
        let checkpoint = match Checkpoint::new(
            plan.name(),
            plan.fingerprint(),
            None,
            sequence,
            state,
            Utc::now(),
        ) {
            Ok(checkpoint) => checkpoint,
            Err(error) => {
                return Err(self.rollback_operation(&transaction, &before, error).await);
            }
        };
        if let Err(error) = self.checkpoints.save(&checkpoint).await {
            return Err(self.rollback_operation(&transaction, &before, error).await);
        }
        self.next_sequence = sequence.checked_add(1);
        Ok(result)
    }

    /// Deletes the durable checkpoint and resets plan and sequence state.
    ///
    /// # Errors
    ///
    /// Returns reset, store, or rollback errors.
    pub async fn reset(&mut self) -> Result<()> {
        let plan = Arc::clone(&self.plan);
        let transaction = plan.leased_transaction(&self.lease).await?;
        let previous_checkpoint = self.checkpoints.load(plan.name()).await?;
        let before = transaction.snapshot().await?;
        if let Err(error) = transaction.reset().await {
            return Err(self.rollback_operation(&transaction, &before, error).await);
        }
        if let Err(error) = self.checkpoints.delete(plan.name()).await {
            let plan_rollback = transaction.restore(&before).await.err();
            let checkpoint_rollback = match &previous_checkpoint {
                Some(checkpoint) => self.checkpoints.save(checkpoint).await.err(),
                None => None,
            };
            let result = compensated_reset_error(error, plan_rollback, checkpoint_rollback);
            if let Some(reason) = &result.poison_reason {
                self.poisoned = Some(reason.clone());
            }
            return Err(result.error);
        }
        self.next_sequence = Some(0);
        self.recovered = true;
        self.poisoned = None;
        Ok(())
    }

    /// Captures plan state through this runner's exclusive lease.
    ///
    /// # Errors
    ///
    /// Returns an operator lifecycle error if state cannot be captured.
    pub async fn plan_snapshot(&self) -> Result<BTreeMap<String, serde_json::Value>> {
        let plan = Arc::clone(&self.plan);
        plan.leased_transaction(&self.lease).await?.snapshot().await
    }

    async fn recover_once(&mut self) -> Result<()> {
        if self.recovered {
            return Ok(());
        }
        let plan = Arc::clone(&self.plan);
        let transaction = plan.leased_transaction(&self.lease).await?;
        if let Some(checkpoint) = self.checkpoints.load(plan.name()).await? {
            validate_checkpoint(&checkpoint, &plan)?;
            let before = transaction.snapshot().await?;
            if let Err(error) = transaction.restore(&checkpoint.state).await {
                return Err(self.rollback_operation(&transaction, &before, error).await);
            }
            self.next_sequence = checkpoint.sequence.checked_add(1);
        }
        self.recovered = true;
        Ok(())
    }

    fn ensure_healthy(&self) -> Result<()> {
        match &self.poisoned {
            Some(reason) => Err(CalcFlowError::Internal {
                message: format!(
                    "streaming runner is poisoned after an incomplete rollback: {reason}; call reset before reuse"
                ),
            }),
            None => Ok(()),
        }
    }

    async fn rollback_operation(
        &mut self,
        transaction: &PlanTransaction<'_>,
        before: &BTreeMap<String, serde_json::Value>,
        original: CalcFlowError,
    ) -> CalcFlowError {
        let result = rollback_error(transaction, before, original).await;
        if let Some(reason) = &result.poison_reason {
            self.poisoned = Some(reason.clone());
        }
        result.error
    }
}
